YarnSpinner/VirtualMachine.cs

```diff
@@ -117,60 +116,209 @@ internal enum TokenType {
+                throw exception;
+            }
+
+            if (newState != null) {
+                this._deserialized = newState;
+            }
 
-        internal class State {
+            return this._deserialized;
+        }
+    }
+
+    public partial class VMStateOption
+    {
+        // Works the same as KeyValuePair, but in protobuf.
 
-            /// The name of the node that we're currently in
-            public string currentNodeName;
+        public VMStateOption(string key, string value) : base()
+        {
+            this.Key = key;
+            this.Value = value;
+        }
+    }
 
-            /// The instruction number in the current node
-            public int programCounter = 0;
+    public partial class VMStateValue
+    {
+        // This entire class exists only because serializing entire Value object
+        // seemed unnecessary so instead we are rebuilding Values from less complex context.
 
-            /// List of options, where each option = <string id, destination node>
-            public List<KeyValuePair<string,string>> currentOptions = new List<KeyValuePair<string, string>>();
+        public VMStateValue(Value yarnValue) : base() {
+            this.YarnValue = yarnValue;
+            this.Type = (int)yarnValue.type;
+            this.Value = yarnValue.AsString;
+        }
 
-            /// The value stack
-            private Stack<Value> stack = new Stack<Value>();
+        private Value _yarnValue;
+        public Value YarnValue {
+            get {
+                if (this._yarnValue == null) {
+                    Yarn.Value.Type type = (Yarn.Value.Type)this.Type;
 
-            /// Methods for working with the stack
-            public void PushValue(object o) {
-                if( o is Value ) {
-                    stack.Push(o as Value);
-                } else {
-                    stack.Push (new Value(o));
+                    switch (type) {
+                        case Yarn.Value.Type.Null:
+                            this._yarnValue = Yarn.Value.NULL;
+                            break;
+                        case Yarn.Value.Type.Number:
+                            this._yarnValue = new Yarn.Value(float.Parse(this.Value));
+                            break;
+                        case Yarn.Value.Type.String:
+                            this._yarnValue = new Yarn.Value(this.Value);
+                            break;
+                        case Yarn.Value.Type.Bool:
+                            this._yarnValue = new Yarn.Value(bool.Parse(this.Value));
+                            break;
+                    }
                 }
+
+                return this._yarnValue;
             }
+            private set {
+                this._yarnValue = value;
+            }
+        }
+    }
 
-            /// Pop a value from the stack
-            public Value PopValue() {
-                return stack.Pop ();
+    public partial class VMState
```

Comment:
Any particular reason why you've renamed the class from State to VMState?

Reply:
Just a naming convention I thought would be better to understand, since it is no longer nested in Virtual machine. Making custom State somewhere doesn't sound as self explanatory as VMState. Should I pull it back?